app/models/volunteer.rb

```diff
@@ -113,35 +113,57 @@ class Volunteer < ApplicationRecord
       .merge(Assignment.inactive)
       .where.not(assignments: { volunteer_id: with_active_assignments.ids })
   }
+
   scope :will_take_more_assignments, (-> { where(take_more_assignments: true) })
-  scope :active, (-> { accepted.with_active_assignments })
 
-  scope :accepted_joined, (-> { accepted.left_outer_joins(:assignments) })
-  scope :loj_without_assignments, (-> { accepted_joined.where(assignments: { id: nil }) })
-  scope :loj_active_take_more, lambda {
-    accepted_joined.will_take_more_assignments.where(assignments: { id: Assignment.active.ids })
+  scope :activeness_not_ended, lambda {
+    where('activeness_might_end IS NULL OR activeness_might_end > ?', Time.zone.today)
+  }
+  scope :activeness_ended, lambda {
+    where(active: true)
+      .where('activeness_might_end IS NOT NULL AND activeness_might_end < ?', Time.zone.today)
+  }
+  scope :active, lambda {
+    activeness_not_ended.where(active: true)
   }
 
+  scope :inactive, lambda {
+    seeking_clients
+  }
   scope :seeking_clients, lambda {
-    accepted_joined
-      .merge(Assignment.inactive)
-      .where.not(assignments: { volunteer_id: with_active_assignments.ids })
-      .or(loj_without_assignments)
+    accepted.where(active: false).or(accepted.activeness_ended)
   }
   scope :seeking_clients_will_take_more, lambda {
-    accepted_joined
-      .merge(Assignment.inactive)
-      .where.not(assignments: { volunteer_id: with_active_assignments.ids })
-      .or(loj_without_assignments)
-      .or(loj_active_take_more)
+    seeking_clients.or(accepted.will_take_more_assignments)
   }
 
+  def verify_and_update_state
+    update(active: active?, activeness_might_end: relevant_period_end_max)
+  end
+
+  def relevant_period_end_max
```

Comment:
There is some implementation of the max dates from the new line 186 of this used in certificates for the computation of duration start and end. If the old implementation is to change then the lines 186-... are not needed and then the duration of the certificates has to be adjusted.

Reply:
Cannot be reused here, because this implementations requires knowledge of nil, and the one you mention filters nil cases

Reply:
Ah, I see now what you mean. I was wrong again. I can use those methods, but as looking at them, I found that they had some flaw, because they where not useing `.pluck(:period_end)`.
So after fixing them, I will make use of these here. Thx